### PR TITLE
Replace QNetworkDiskCache with custom cache

### PR DESCRIPTION
**Pull request by ad-daniel**
**Description**
This PR introduces a different caching system for assets. Images, meshes, sounds (and eventually PROTO) are downloaded on the fly and saved in the `.cache/Cyberbotics/assets/` folder. The cache is opaque but deterministic, meaning each remote URL corresponds to a unique local file (identified by the SHA1 of the URL). As such, the assets folder can be:
- distributed as a zip file, allowing for offline usage of Webots and faster loading in webots.cloud
- the size of the distribution can be reduced further (PROTO will be defined as EXTERNPROTO references for their download)
- as the URLs refer to a specific release (defined by a tag), support for different versions of these assets (most notably PROTO) can be ensured.

Maximal cache size is defined but not strictly enforced (not without a restart). Meaning if the max size is defined to 512 MB, and loading a world results in this size being exceeded (for instance 600MB), it will require restarting Webots for the excess to be purged (by removing first the least used ones (oldest `lastRead`). 

Cached files are:
- [x] Backgrounds
- [x] Textures
- [x] Camera noise mask
- [x] Sounds
   - [x] Motor
   - [x] ContactProperties
   - [x] Speaker
- [x] Meshes 
- [x] Add test in test suite
   - [x] Textures
   - [x] Meshes
- [x] Cleanup/Testing

### CODE REVIEW

File: src/webots/core/WbNetwork.hpp
Comment (omichel) [inline src/webots/core/WbNetwork.hpp:45]: Why can't we do that?
```suggestion
  static const QString urlToHash(const QString &url) const;
```

File: src/webots/gui/WbPreferencesDialog.cpp
Comment (omichel) [inline src/webots/gui/WbPreferencesDialog.cpp:199]:
```suggestion
    WbMessageBox::info(tr("For optimal performance, at least\n512 MB of cache are necessary."), this);
```

File: src/webots/nodes/WbContactProperties.cpp
Comment (omichel) [inline src/webots/nodes/WbContactProperties.cpp:231]: This TODO should be addressed.

File: src/webots/nodes/WbContactProperties.cpp
Comment (omichel) [inline src/webots/nodes/WbContactProperties.cpp:24]: This enum seems unused.
```suggestion
```

File: src/webots/nodes/WbImageTexture.cpp
Comment (omichel) [inline src/webots/nodes/WbImageTexture.cpp:143]:
```suggestion
  const QString filePath = isWebAsset ? WbNetwork::instance()->get(url) : path(true);
```

File: src/webots/nodes/WbMotor.cpp
Comment (omichel) [inline src/webots/nodes/WbMotor.cpp:38]: Seems useless:
```suggestion
```

File: src/webots/nodes/utils/WbDownloader.hpp
Comment (omichel) [inline src/webots/nodes/utils/WbDownloader.hpp:19]: That seems useless:
```suggestion
```

File: src/webots/nodes/utils/WbDownloader.cpp
Comment (omichel) [inline src/webots/nodes/utils/WbDownloader.cpp:18]: We should probably keep it here and not in the header file.

File: tests/api/controllers/cache_checker/cache_checker.c
Comment (omichel) [inline tests/api/controllers/cache_checker/cache_checker.c:37]:
```suggestion
                              "In test %d: wrong mesh %d is applied (distance is %f instead of %f)", test_number, i, distance, expected_distance);
```

**Review by omichel — CHANGES_REQUESTED** [review]
This is a first code review.
It looks overall good to me.
Only the documentation should be updated in a different way.

**Comment by ad-daniel** [comment]
Looks like removing the `md` files from `update_urls` makes it unnecessary to force the encoding

**Review by omichel — DISMISSED** [review]
Since it is a quite big PR, it would be nice to have a second review.

File: src/webots/core/WbNetwork.cpp
Comment (BenjaminDeleze) [inline src/webots/core/WbNetwork.cpp:130]: Is it possible to reach this code ? 
The `assert(isCached(url))` return true only if `cacheMap` already contains the url or if the file exists on the disk and in the latest case, it inserts it in the `cacheMap`.
So if we pass the assert, the test at line 124 will always pass?

**Review by BenjaminDeleze — CHANGES_REQUESTED** [review]
I managed to crash webots:

- Open `nao_room.wbt`
- Go in `Worldinfo` -> `contactProperties`
- Select one `ContactProperties` and change one of its urls to an invalid one
- Webots crash.

Error message: ```Warning: QIODevice::read (QFile, "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/default/worklds/sounds/slide.wav"): device not open
terminate called after throwing an instance of 'char const*'
./webots: line 96: 115567 Aborted                 (core dumped) "$webots_home/bin/webots-bin" "$@"```

File: src/webots/core/WbNetwork.cpp
Comment (ad-daniel) [reply src/webots/core/WbNetwork.cpp:130]: The assert is a sanity check when developing. If it's triggered it means we didn't implement the retrieval logic correctly, in other words we shouldn't have reached to point of calling the get function unless the file is in the local cache.

`cacheMap` is not a complete representation of the state of the cache, it's a representation that we build progressively as assets are requested/returned, and the information is lost when closing Webots. 
By keeping this internal representation allows to skip having to compute the hash from the url and checking if the file exists on the disk every time, which otherwise we would do even when the same url was requested before (or rather, in the current session).

**Comment by omichel** [comment]
When clearing the cache from the Preferences window, Webots doesn't close the Preference window. When changing the cache size, it closes the Preferences window. This is not consistent. I believe changing the cache size should not close the Preferences window.

**Comment by omichel** [comment]
Also, when opening the Preferences windows, doing not and pressing the "OK" button, we get a message window saying the cache size was updated. This should not happen.

**Review by BenjaminDeleze — CHANGES_REQUESTED** [review]

**Comment by BenjaminDeleze** [comment]
> I managed to crash webots:
> 
> * Open `nao_room.wbt`
> * Go in `Worldinfo` -> `contactProperties`
> * Select one `ContactProperties` and change one of its urls to an invalid one
> * Webots crash.
> 
> Error message: `Warning: QIODevice::read (QFile, "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/default/worklds/sounds/slide.wav"): device not open terminate called after throwing an instance of 'char const*' ./webots: line 96: 115567 Aborted (core dumped) "$webots_home/bin/webots-bin" "$@"`

It crashes also when changing the url of the sound of a motor

**Review by BenjaminDeleze — CHANGES_REQUESTED** [review]
I found another crash (in a very rare edge case):
- Load a world.
- Clear the cash.
- Reload the world.

**Comment by ad-daniel** [comment]
All issues should be addressed now, both with and without internet connection

**Review by BenjaminDeleze — DISMISSED** [review]
Thank you!

I tried to make it bug in a lot of different ways but did not find any strange behaviors

File: src/webots/core/WbNetwork.cpp
Comment (omichel) [inline src/webots/core/WbNetwork.cpp:31]: We usually name these static variable with a `g` prefix.

File: src/webots/gui/WbPreferencesDialog.cpp
Comment (omichel) [inline src/webots/gui/WbPreferencesDialog.cpp:516]:
```suggestion
    new QLabel(tr("Amount of cache used: %1 MB.").arg(WbNetwork::instance()->cacheSize() / (1024 * 1024)), this), 1, 0);
```

File: src/webots/nodes/WbCamera.cpp
Comment (omichel) [inline src/webots/nodes/WbCamera.cpp:161]: Why can't we keep it const ref?

File: src/webots/nodes/WbImageTexture.cpp
Comment (omichel) [inline src/webots/nodes/WbImageTexture.cpp:143]:
```suggestion
    warn(tr("Texture file could not be read: '%1'").arg(filePath));
```

File: src/webots/nodes/utils/WbDownloader.cpp
Comment (omichel) [inline src/webots/nodes/utils/WbDownloader.cpp:97]:
```suggestion
    mError = tr("Cannot download '%1', error code: %2: %3")
```

**Review by omichel — CHANGES_REQUESTED** [review]
A few requests...

File: src/webots/nodes/WbMesh.cpp
Comment (omichel) [inline src/webots/nodes/WbMesh.cpp:147]:
```suggestion
      warn(tr("Mesh file could not be read: '%1'").arg(filePath));
```

File: src/webots/wren/WbWrenCamera.cpp
Comment (omichel) [inline src/webots/wren/WbWrenCamera.cpp:348]:
```suggestion
      return tr("Cannot load '%1': %2").arg(noiseMaskUrl).arg(imageReader->errorString());
```

**Review by omichel — CHANGES_REQUESTED** [review]

File: src/webots/nodes/WbCamera.cpp
Comment (ad-daniel) [reply src/webots/nodes/WbCamera.cpp]: my bad, it can and should be

**Review by omichel — APPROVED** [review]
Thank you.

**Pull request merged**
